File: common/ocaml_common/text_manipulate.ml
(* ------------------------------------------------------------------- *)
(*
	text_manipulate.ml

					Jul/25/2012

*)
(* ------------------------------------------------------------------- *)
open	Printf
open	Str
open	Unix
(* ------------------------------------------------------------------- *)
type unit_template = {name:string; population:int; date_mod:string}


(* ------------------------------------------------------------------- *)
let dict_append_proc = fun dict_aa key name_in population_in date_mod_in ->
	Hashtbl.add dict_aa key
	{name= name_in; population= population_in; date_mod= date_mod_in}
(* ------------------------------------------------------------------- *)
let text_read_proc_exec = fun file_text dict_aa delimit ->
	let ic = open_in file_text in
	try while true do
		let cols = (split (regexp delimit) (input_line ic))
		in
		dict_append_proc dict_aa (List.nth cols 0)
			 (List.nth cols 1)
			 (int_of_string (List.nth cols 2))
			 (List.nth cols 3)
	done with End_of_file ->
		close_in ic;
	print_endline "*** text_read_proc *** ppp ***";
	print_endline "*** text_read_proc *** qqq ***"

(* ------------------------------------------------------------------- *)
let text_read_proc = fun file_text dict_aa ->
	let delimit = "[\t]+"
	in
	text_read_proc_exec file_text dict_aa delimit
(* ------------------------------------------------------------------- *)
let join_proc = fun key uu delimit ->
	key ^ delimit ^ uu.name ^ delimit ^ (string_of_int uu.population)
		^ delimit ^ uu.date_mod
(* ------------------------------------------------------------------- *)
let out_str_proc  = fun oc delimit key_aa unit_aa ->
	fprintf oc "%s\n" (join_proc key_aa unit_aa delimit)
(* ------------------------------------------------------------------- *)
let text_write_proc_exec = fun text_file dict_aa delimit ->
	let oc = open_out text_file in
		Hashtbl.iter (fun key value ->
			out_str_proc oc delimit key value) dict_aa;
		close_out oc

(* ------------------------------------------------------------------- *)
let text_write_proc = fun text_file dict_aa ->
	let delimit = "\t"
	in text_write_proc_exec text_file dict_aa delimit
(* ------------------------------------------------------------------- *)
let record_display_proc = fun key unit_aa ->
	printf "%s\t" key;
	printf "%s\t" unit_aa.name;
	printf "%d\t" unit_aa.population;
	printf "%s\n" unit_aa.date_mod


(* ------------------------------------------------------------------- *)
let dict_display_proc = fun dict_aa ->
	Hashtbl.iter record_display_proc dict_aa
(* ------------------------------------------------------------------- *)
(*
let get_current_date_proc = fun () ->
	let t = Unix.localtime (Unix.time ())
	in let (day, month, year) = (t.tm_mday, t.tm_mon, t.tm_year)
	in let today = string_of_int (1900 + year) ^ "-" ^ string_of_int (month + 1)
	^ "-" ^string_of_int day
	in today
*)
let get_current_date_proc = 
		let t = Unix.localtime (Unix.time ())
	in
		let (day, month, year) = (t.tm_mday, t.tm_mon, t.tm_year)
	in
		 string_of_int (1900 + year) ^ "-" ^
			string_of_int (month + 1) ^ "-" ^string_of_int day

(* ------------------------------------------------------------------- *)
let dict_update_proc = fun dict_aa key population_in ->
	let _ =	printf "%s\n" key
	in let _ =	printf "%d\n" population_in
	in let unit_aa = Hashtbl.find dict_aa key
	in let today = get_current_date_proc
	in let unit_new = {name= unit_aa.name; population= population_in; date_mod= today}

	in let _ = Hashtbl.replace dict_aa key unit_new
	in
	0
(* ------------------------------------------------------------------- *)
let csv_read_proc = fun file_text dict_aa ->
	let delimit = "[,]+"
	in
	text_read_proc_exec file_text dict_aa delimit
(* ------------------------------------------------------------------- *)
let csv_write_proc = fun text_file dict_aa ->
	let delimit = ","
	in text_write_proc_exec text_file dict_aa delimit
(* ------------------------------------------------------------------- *)
(* ------------------------------------------------------------------- *)
let rec file_to_str_proc filename =
	let stat = Unix.stat filename in
	let size = stat.Unix.st_size in
	let ii = open_in filename in
	let buffer = String.make size '\000' in
	begin
		ignore (really_input ii buffer 0 size);
		close_in ii;
		buffer
	end
(* ------------------------------------------------------------------- *)

File: common/go_common/text_manipulate.go
// ----------------------------------------------------------------
//
//	text_manipulate.go
//
//					May/20/2018
//
// ----------------------------------------------------------------
// package text_manipulate
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"strconv"
	"time"
	"sort"
	"reflect"
)

// ----------------------------------------------------------------
func display_record_proc (key string, unit_aa map[string]interface{}) {
	fmt.Printf ("%s\t",key)
	fmt.Printf ("%s\t",unit_aa["name"])
	vv := reflect.TypeOf(unit_aa["population"])
//	fmt.Println(v)

	population := 0
	if (vv.Kind() == reflect.Int) {
		population = unit_aa["population"].(int)
		} else {
	pf := unit_aa["population"].(float64)
	population = int(pf)
		}

	str_population := strconv.Itoa(population)
	fmt.Printf ("%s\t",str_population)
	fmt.Printf ("%s\n",unit_aa["date_mod"])
}

// ----------------------------------------------------------------
func dict_display_proc (dict_aa map[string](map[string]interface{})) {
	var keys []string
	for key,_ := range dict_aa {
	keys = append (keys,key)
	}

	sort.Strings (keys)

	for _, key := range keys {
	display_record_proc (key, dict_aa[key])
	}

}

// ----------------------------------------------------------------
func unit_gen_proc (name string,population int,date_mod string) map[string]interface{} {
	unit_aa := make (map[string]interface{})
	unit_aa["name"] = name
	unit_aa["population"] = population
	unit_aa["date_mod"] = date_mod

	return (unit_aa)
}

// ----------------------------------------------------------------
// [4-4-6]:
func join_proc (id,name,population,date_mod,delimit string) string {
	arry := []string {id,name,population,date_mod}

	datax := strings.Join (arry,delimit)
	datax += "\n"

	return (datax)
}

// ----------------------------------------------------------------
// [4-4]:
func dict_to_string_proc (dict_aa map[string](map[string]interface{}),delimit string) string {
	str_out := ""
	for key,value := range dict_aa {
	population := value["population"].(int)
	str_population := strconv.Itoa(population)
	str_out += join_proc (key,value["name"].(string),
		str_population,value["date_mod"].(string),delimit)
	}

	return	(str_out)
}

// ----------------------------------------------------------------
// [4]:
func text_write_proc (text_file string,dict_aa map[string](map[string]interface{})){
	delimit := "\t"
	str_out := dict_to_string_proc (dict_aa,delimit)
	ioutil.WriteFile (text_file, []byte(str_out), 0666)
}

// ----------------------------------------------------------------
// [6-4]:
func text_read_proc_exec (text_file,delimit string) map[string](map[string]interface{}){
	dict_aa := make (map[string](map[string]interface{}))


	buff,_ := ioutil.ReadFile(text_file)

//	datax := strings.Split (string (buff),"\n",-1)
	datax := strings.Split (string (buff),"\n")

	fmt.Printf ("%d\n" , len (datax))

	for _, pp := range datax {
//		data_row := strings.Split (pp,delimit,-1)
		data_row := strings.Split (pp,delimit)
		if (1 < (len (data_row))) {
		key := data_row[0]
		str_population := data_row[2]
		population,_ := strconv.Atoi(str_population)
		dict_aa[key] = unit_gen_proc (data_row[1],population,data_row[3])
			}
		}

	fmt.Printf ("%d\n" , len (dict_aa))
	return (dict_aa)
}

// ----------------------------------------------------------------
// [6]:
func text_read_proc (text_file string) map[string](map[string]interface{}){
	delimit := "\t"
	dict_aa := text_read_proc_exec (text_file,delimit)

	return (dict_aa)
}

// ----------------------------------------------------------------
func get_current_date_proc () string {
	now := time.Now ()
	fmt.Printf ("%s\t" ,now)
	fmt.Printf ("%d-%d-%d\n" ,now.Year (),now.Month(),now.Day())
	today := strconv.Itoa (now.Year()) + "-" +
		fmt.Sprintf ("%d",now.Month()) + "-" +
		strconv.Itoa (now.Day())

	return	today
}

// ----------------------------------------------------------------
func dict_update_proc (dict_aa map[string](map[string]interface{}),
	key string,population_in int) map[string](map[string]interface{}) {

	fmt.Printf ("key = %s\n", key)

	name_aa := dict_aa[key]["name"].(string)
	unit_aa := make (map[string]interface{})
	unit_aa["name"] = name_aa
	unit_aa["population"] = population_in
	unit_aa["date_mod"] = get_current_date_proc ()

	dict_aa[key] = unit_aa

	return (dict_aa)
}

// ----------------------------------------------------------------
// [8]:
func csv_write_proc (text_file string,dict_aa map[string](map[string]interface{})){
	delimit := ","
	str_out := dict_to_string_proc (dict_aa,delimit)
	ioutil.WriteFile (text_file, []byte(str_out), 0666)
}

// ----------------------------------------------------------------
// [10]:
func csv_read_proc (text_file string) map[string](map[string]interface{}){
	delimit := ","

	dict_aa := text_read_proc_exec (text_file,delimit)
	return (dict_aa)
}

// ----------------------------------------------------------------
